Lib/ufo2ft/instructionCompiler.py

```diff
@@ -0,0 +1,267 @@
+        self._check_tt_data_format(ttdata, f"glyph '{glyph.name}'")
+        glyph_hash = ttdata.get("id", None)
+        if not self._check_glyph_hash(glyph, ttglyph, glyph_hash):
+            return
+
+        # Compile the glyph program
+        asm = ttdata.get("assembly", None)
+        if asm is None:
+            # The "assembly" key is required.
+            logger.error(
+                f"Glyph assembly missing, glyph '{glyph.name}' will have "
+                "no instructions in font."
+            )
+            return
+
+        if not asm:
+            # If the assembly code is empty, don't bother adding a program
+            logger.debug(f"Glyph '{glyph.name}' has no instructions.")
+            return
+
+        ttglyph.program = ttLib.tables.ttProgram.Program()
+        ttglyph.program.fromAssembly(asm)
+
+    def _set_composite_flags(self, glyph, ttglyph):
+        # Set component flags
+
+        if len(ttglyph.components) != len(glyph.components):
+            # May happen if nested components have been flattened by a filter
+            logger.error(
+                "Number of components differ between UFO and TTF "
+                f"in glyph '{glyph.name}' ({len(glyph.components)} vs. "
+                f"{len(ttglyph.components)}, not setting component flags from"
+                "UFO. They may still be set heuristically."
+            )
+            return
+
+        # We need to decide when to set the flags.
+        # Let's assume if any lib key is not there, or the component
+        # doesn't have an identifier, we should leave the flags alone.
+
+        # Keep track of which component has the USE_MY_METRICS flag
+        use_my_metrics_comp = None
+
+        for i, c in enumerate(ttglyph.components):
+            ufo_component_id = glyph.components[i].identifier
+            if ufo_component_id is None:
+                # No information about component flags is stored in the UFO,
+                # use heuristics.
+
+                # https://github.com/googlefonts/ufo2ft/pull/425 recommends
+                # to always set the ROUND_XY_TO_GRID flag
+                c.flags |= ROUND_XY_TO_GRID
+            elif (
+                OBJECT_LIBS_KEY in glyph.lib
+                and ufo_component_id in glyph.lib[OBJECT_LIBS_KEY]
+                and (
+                    TRUETYPE_ROUND_KEY in glyph.lib[OBJECT_LIBS_KEY][ufo_component_id]
+                    or TRUETYPE_METRICS_KEY
+                    in glyph.lib[OBJECT_LIBS_KEY][ufo_component_id]
+                )
+            ):
+                component_lib = glyph.lib[OBJECT_LIBS_KEY][ufo_component_id]
+
+                # https://github.com/googlefonts/ufo2ft/pull/425 recommends
+                # to always set the ROUND_XY_TO_GRID flag, so we only
+                # unset it if explicitly done so in the lib
+                if component_lib.get(TRUETYPE_ROUND_KEY, True):
+                    logger.info("    ROUND_XY_TO_GRID")
```

Comment:
does this emit a logging INFO by default (if TRUETYPE_ROUND_KEY is missing) _for every component_ in the font?!

Reply:
Good point, we don’t want that.